RecoLocalCalo/HGCalRecProducers/plugins/alpaka/HGCalLayerClustersSoAAlgoWrapper.dev.cc

```diff
@@ -0,0 +1,228 @@
+
+  using namespace cms::alpakatools;
+
+  // Set energy and number of hits in each clusters
+  class HGCalLayerClustersSoAAlgoKernelEnergy {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                  const unsigned int numer_of_clusters,
+                                  const HGCalSoARecHitsDeviceCollection::ConstView input_rechits_soa,
+                                  const HGCalSoARecHitsExtraDeviceCollection::ConstView input_clusters_soa,
+                                  HGCalSoAClustersDeviceCollection::View outputs) const {
+      constexpr const int kInvalidCluster = -1;
+      // make a strided loop over the kernel grid, covering up to "size" elements
+      for (int32_t i : uniform_elements(acc, input_rechits_soa.metadata().size())) {
+        // Skip unassigned rechits
+        if (input_clusters_soa[i].clusterIndex() == kInvalidCluster) {
+          continue;
+        }
+        auto clIdx = input_clusters_soa[i].clusterIndex();
+        alpaka::atomicAdd(acc, &outputs[clIdx].energy(), input_rechits_soa[i].weight());
+        alpaka::atomicAdd(acc, &outputs[clIdx].cells(), 1);
+        if (input_clusters_soa[i].isSeed() == 1) {
+          outputs[clIdx].seed() = input_rechits_soa[i].detid();
+        }
+      }
+    }
+  };
+
+  // Kernel to find the max for every cluster
+  class HGCalLayerClustersSoAAlgoKernelPositionByHits {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                  const unsigned int numer_of_clusters,
+                                  float thresholdW0,
+                                  float positionDeltaRho2,
+                                  const HGCalSoARecHitsDeviceCollection::ConstView input_rechits_soa,
+                                  const HGCalSoARecHitsExtraDeviceCollection::ConstView input_clusters_soa,
+                                  HGCalSoAClustersDeviceCollection::View outputs,
+                                  HGCalSoAClustersExtraDeviceCollection::View outputs_service) const {
+      // make a strided loop over the kernel grid, covering up to "size" elements
+      for (int32_t hit_index : uniform_elements(acc, input_rechits_soa.metadata().size())) {
+        const int cluster_index = input_clusters_soa[hit_index].clusterIndex();
+
+        // Bail out if you are not part of any cluster
+        if (cluster_index == -1) {
+          continue;
+        }
+
+        alpaka::atomicAdd(acc, &outputs_service[cluster_index].total_weight(), input_rechits_soa[hit_index].weight());
+        // Read the current seed index, and the associated energy.
+        int clusterSeed = outputs_service[cluster_index].maxEnergyIndex();
+        float clusterEnergy = (clusterSeed == -1) ? 0.f : input_rechits_soa[clusterSeed].weight();
```

Comment:
```suggestion
        float clusterEnergy = (clusterSeed == maxEnergyIndex) ? 0.f : input_rechits_soa[clusterSeed].weight();
```
?

Or is this a different `-1` ?

Reply:
This is a logically different -1. Well, in reality -1 is just used as in invalid index, but the meaning is logically different. That's why I defined 2 different constants in name, identical in value.